### PR TITLE
Integrate detector for deteriorated javadoc

### PR DESCRIPTION
**Pull request by rsmidt**
This pull request will integrate the work of Michel K. into the tools ecosystem. It's now possible to detect faulty or deteriorated javadoc (e.g. missing tags, descriptions etc.).

In order to achieve that we had to upgrade the spoon dependency to v8.0.0 (at least 7.5.0) and also stub some collections from googles guava library to avoid a dependency.

Michels work is untouched except some migrations in his tests to the latest version of libvcs4j and some proper formatting (IDEAs default).

### CODE REVIEW

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/SpoonModel.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package de.unibremen.informatik.st.libvcs4j.spoon;
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/SpoonModel.java]: This class was renamed to [Environment](https://github.com/uni-bremen-agst/libvcs4j/blob/master/libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/Environment.java). Please remove it.

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/SpoonModelBuilder.java [hunk 2]
```diff
@@ -0,0 +1,459 @@
+package de.unibremen.informatik.st.libvcs4j.spoon;
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/SpoonModelBuilder.java]: This class was renamed to [EnvironmentBuilder](https://github.com/uni-bremen-agst/libvcs4j/blob/master/libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/EnvironmentBuilder.java). Please remove it.

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     * Tries to put a value into the sink associated with a given key.
+     *
+     * @param key   the key of the sink.
+     * @param value the value to insert.
+     * @return {@code true} if the the element was inserted, otherwise {@code false}.
+     */
+    @Override
+    public boolean put(final K key, final V value) {
+        var coll = map.computeIfAbsent(key, k -> new ArrayList<>());
+
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java]: Unnecessary new line

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                .collect(Collectors.toList()));
+    }
+
+    private Stream<K> toKeys(Map.Entry<K, ArrayList<V>> entry) {
+        return entry.getValue().stream().map(e -> entry.getKey());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java:81]: I think you can simply delegate `o` to `map`.

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
+        ArrayListMultimap<?, ?> that = (ArrayListMultimap<?, ?>) o;
+        return Objects.equals(map, that.map);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(map);
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java]: I think you can simply delegate this call to `map`.

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultiset.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+     * @param value the value which occurrences have to be counted.
+     */
+    @Override
+    public int count(final V value) {
+        return (int) values.stream().filter(v -> v.equals(value)).count();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultiset.java:106]: See ArrayListMultimap.

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultiset.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
+        ArrayListMultiset<?> that = (ArrayListMultiset<?>) o;
+        return Objects.equals(values, that.values);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(values);
```
Comment (msteinbeck) [inline libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultiset.java]: See ArrayListMultimap

File: libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.List;
+
+import static org.junit.Assert.*;
+
+public class ArrayListMultimapTest {
+
+    @Test
+    public void create() {
+        final var actual = ArrayListMultimap.create();
+
```
Comment (msteinbeck) [inline libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java:15]: Unnecessary new line.

File: libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        given.put("key2", "item2");
+
+        final var values = List.of("key1", "key1", "key2");
+        final var expected = new ArrayListMultiset<>(values);
+
+        final var actual = given.keys();
+
+        assertEquals(expected, actual);
+    }
+}
```
Comment (msteinbeck) [inline libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java]: Missing new line at end of file.

File: libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultisetTest.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+    @Test
+    public void count() {
+        final var given = new ArrayListMultiset<>(List.of("item1", "item1", "item2", "item3"));
+
+        assertEquals(2, given.count("item1"));
+        assertEquals(1, given.count("item2"));
+        assertEquals(1, given.count("item3"));
+    }
+}
```
Comment (msteinbeck) [inline libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultisetTest.java]: Missing new line at end of file.

**Comment by msteinbeck** [comment]
Several of your files are missing a new line character at EOF. Please add them.

**Comment by msteinbeck** [comment]
Travis fails on OSX because the image uses JDK8. We have to upgrade it to JDK11 in order to compile and run the tests.

File: libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java (hunk 8 above)
Comment (rsmidt) [reply libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java:15]: Do you have a style guide document declaring those rules?

**Comment by rsmidt** [comment]
> Travis fails on OSX because the image uses JDK8. We have to upgrade it to JDK11 in order to compile and run the tests.

According to https://docs.travis-ci.com/user/reference/osx/ the xcode10.1 image already includes jdk 11. Should I try it with the newest image version?

**Comment by rsmidt** [comment]
> Several of your files are missing a new line character at EOF. Please add them.

I'll add them everywhere except for the test resources. Adding newlines there will break the tests.

File: libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java (hunk 4 above)
Comment (rsmidt) [reply libvcs4j-tools/src/main/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimap.java:81]: If you mean to directly return `values.equals(o)`, that won't work. We first have to assert that o is of type ArrayListMultimap.

File: libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java (hunk 8 above)
Comment (msteinbeck) [reply libvcs4j-tools/src/test/java/de/unibremen/informatik/st/libvcs4j/spoon/codesmell/missingJavadoc/ArrayListMultimapTest.java:15]: We neither have a strict style guide nor do we use a style checker. It is the responsibility of the developers to use a "sane" style. That being said, I prefer:

1. Tabs for indention, spaces for layout (1 tab = 4 spaces)
2. Maximum width of 80 (exclusive) spaces.
3. Avoid useless new lines.

**Comment by msteinbeck** [comment]
> I'll add them everywhere except for the test resources. Adding newlines there will break the tests.

That's somehow unexpected. However, if they are necessary to keep tests intact, just leave them.

**Comment by msteinbeck** [comment]
> According to https://docs.travis-ci.com/user/reference/osx/ the xcode10.1 image already includes jdk 11. Should I try it with the newest image version?

Actually, the image is using JDK13 (from Travis log):

```
$ java -Xmx32m -version
openjdk version "13.0.1" 2019-10-15
OpenJDK Runtime Environment (build 13.0.1+9)
OpenJDK 64-Bit Server VM (build 13.0.1+9, mixed mode, sharing)
$ javac -J-Xmx32m -version
javac 13.0.1
23.29s$ ./gradlew assemble
Downloading https://services.gradle.org/distributions/gradle-5.2.1-all.zip
````
It looks like Gradle 5.2.1 isn't compatible with JDK13. I guess you have to add `jdk: openjdk11` to [.travis.yml](https://github.com/uni-bremen-agst/libvcs4j/blob/master/.travis.yml#L8)

**Comment by msteinbeck** [comment]
I'm wondering why `de.unibremen.informatik.st.libvcs4j.spoon.EnvironmentBuilderGSONTest > nullPointerInUpdateFrom4To6` fails. The code smell detector introduced by this commit is not related to this test. I guess it results from upgrading spoon to 8.0.0. I'll try to figure out what is going wrong here.

**Comment by rsmidt** [comment]
> > According to https://docs.travis-ci.com/user/reference/osx/ the xcode10.1 image already includes jdk 11. Should I try it with the newest image version?
> 
> Actually, the image is using JDK13 (from Travis log):
> 
> ```
> $ java -Xmx32m -version
> openjdk version "13.0.1" 2019-10-15
> OpenJDK Runtime Environment (build 13.0.1+9)
> OpenJDK 64-Bit Server VM (build 13.0.1+9, mixed mode, sharing)
> $ javac -J-Xmx32m -version
> javac 13.0.1
> 23.29s$ ./gradlew assemble
> Downloading https://services.gradle.org/distributions/gradle-5.2.1-all.zip
> ```
> 
> It looks like Gradle 5.2.1 isn't compatible with JDK13. I guess you have to add `jdk: openjdk11` to [.travis.yml](https://github.com/uni-bremen-agst/libvcs4j/blob/master/.travis.yml#L8)

I locked the version to openjdk11. Now it builds but some test fail, both under linux and mac os.

**Pull request merged**